[WIP] Mdd rds split fix

michael-dev2rights opened the pull request:
Here are some proposed fixes which got me further through the integration tests that are attached.

a) making instance state a function so that it updates whilst we are waiting for it to change
b) handling various cases where I got exceptions from nonexistent parameters
c) handling other parameter situations

I'm just learning my way around the code so these may be a bit the wrong thing, however hopefully they are useful.   I can try to write up some of the problems I had into this ticket to justify this but it will take a little while before I can do so.

mikedlr committed:
rds_instance module - fix various parameter handling cases from trying to run integration test cases

willthames commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -572,7 +582,7 @@ def modify_db_instance(module, conn):
     # guess that this changed the DB, need a way to check
     diff = existing_instance.diff(instance.data)
-    changed = not not diff
+    changed = not diff
```
`diff == None` => `not diff == True` => `changed = True`

which is wrong. If diff is empty, changed should be False (using `not not` to convert it to a boolean)

mikedlr replied:
I was wondering, However the interesting thing was that this change fixed the integration test case which was meant to show change and which had a positive existing diff.

willthames commented on lib/ansible/module_utils/rds.py:
```diff
@@ -51,16 +51,18 @@ def __init__(self, dbinstance):
         self.data = self.get_data()
 
+    def status(self):
+        return self.instance.get('DBInstanceStatus')
```
self.instance is really a dict, isn't it -  making this a function won't have any real effect, as far as I can tell - I could be wrong though.

mikedlr replied:
the call to self.instance.get was already a function.  Before this was an instance variable so the function would be called only once during object initialisation.  As it is now, whenever status is called self.instance.get('DBInstanceStatus') should be called as a function which I think _should_ help

willthames commented:
This needs rebasing to reflect latest changes (in particular, ansible.module_utils.rds has had some major changes from #3)

mikedlr committed:
rds_instance module, rds in mod utils and int tests - small fixes to make tests go

mikedlr committed:
rds_instance module - refresh instance in case of state just before waiting - seems to fix test

michael-dev2rights commented:
@willthames I have rebased, added more fixes and removed the "not not" change. 

Removing the not not breaks "check modify instance changed" because "modify rds instance" returns changed=false even though the diff returned actually shows that it _has_ changed the instance port number.  Something must be wrong in typecasting the diff to a boolean.

willthames commented:
Thanks @michael-dev2rights 

I don't really understand why the `not not diff` isn't working, my cursory checking of this suggests it should:

```
>>> diff = dict()  # empty diff, no change => changed=False
>>> not not diff
False
>>> diff = dict(hello='world') # non-empty diff => changed=True
>>> not not diff
True
```

I'll try and find some time to make some more progress on this soon.

mikedlr committed:
Add some simple unit tests of the diff functionality - failing at this time to show problems

mikedlr committed:
rds_instance and related modules - Varied fixes to get more integration tests working

michael-dev2rights commented:
@willthames I checked it again and in the end it seems the `not not` is correct but the wrong instance object was being checked or something similar.  I am playing around with this.  I've pushed up something that won't be right but will show what I've been working on.

mikedlr committed:
rds_instance - fixes & tests to ensure wait instance and changed state work

mikedlr committed:
rds_instance - port parameter is optional so ignore if missing

mikedlr committed:
rds_instance integration tests - fix non modify and disable replica promition

mikedlr committed:
rds_instance module - full traceback of exceptions in fail_json better error reporting

willthames commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -429,7 +441,8 @@ def replicate_db_instance(module, conn):
         except botocore.exceptions.ClientError as e:
-            module.fail_json(msg="Failed to create replica instance: %s " % str(e))
+            module.fail_json(msg="Failed to create replica instance: %s " % str(e),
+                             exception=traceback.format_exc() )
```
boto3 exception guidelines also say to add `**camel_dict_to_snake_dict(e.response)`

mikedlr replied:
should now be fixed.

willthames commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -353,19 +356,28 @@
-    while wait_timeout > time.time() and resource.status != status:
+    # Refresh the resource immediately in case we just changed it's state; should we sleep first?
+    resource = get_db_instance(conn, resource.name)
+    eprint(str(resource))
```
You can also use the `q` module for this.

michael-dev2rights replied:
I had a look at q; it's fine for debugging and I'll probably use it somehow, but it seems that it could leave debugging information in a public file in /tmp which could be a security problem if it accidentally went into production so I'm not totally convinced by it and I won't use it in any case in code I push up to github until I better understand how to use it securely.

willthames commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -441,7 +454,15 @@ def replicate_db_instance(module, conn):
-    valid_vars = ['snapshot', 'skip_final_snapshot']
+
+    #we have to accept but ignore variables which have defaults
+    valid_vars = ['snapshot', 'skip_final_snapshot', 'storage_type']
```
For `storage_type`, I was thinking of removing the default from the argument_spec, and letting the AWS API handle the default. This would mean updating the `diff` check to handle storage_type though (probably using your `leave_if_null` implementation)

willthames commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -696,56 +748,63 @@ def validate_parameters(required_vars, valid_vars, module):
-            force_failover = dict(type='bool', default=False),
-            force_password_update = dict(type='bool', default=False),
-        )
+argument_spec = ec2_argument_spec()
```
Why has this moved outside of main?

michael-dev2rights replied:
I use these definitions as part ow what's needed to seed Mock objects in the tests.  That means I need a reference to them.  It seems to me that this is reasonable ; I don't see any need to hide them though I would love to shrink the code.

willthames commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -14,6 +14,9 @@
 # along with Ansible. If not, see <http://www.gnu.org/licenses/>.
 
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
```
Are we adding this just for `eprint` - if so, just use `q` instead.

michael-dev2rights replied:
I'm putting it in generally as a cut and paste understanding it should reduce compatibility problems by making more python 3 code behave the same as python 2 more of the time.

willthames commented on test/units/modules/cloud/amazon/test_rds.py:
```diff
@@ -18,12 +18,115 @@
+import ansible.module_utils.rds as rds_u
 
+from ansible.module_utils.basic import AnsibleModule
+import ansible.module_utils.basic as ba
```
No need for this shortening.

willthames commented on test/units/modules/cloud/amazon/test_rds.py:
```diff
@@ -18,12 +18,115 @@
+
+def test_modify_should_return_false_in_changed_if_param_same():
+    ba._ANSIBLE_ARGS='{"ANSIBLE_MODULE_ARGS":{"instance_name":"fred", "port": 342}}'
+    params={"port":342,"force_password_update":True,"instance_name":"fred"}
```
test code should also generally pass pep8 - this dict definition is unnecessarily terse.

mikedlr replied:
my flake8 is ignoring this for some reason only it knows and my bad old C programmer habits are hard to break.  Hopefully improved now but I will fix any more I spot.

mikedlr committed:
rds_instance - add import for traceback missing from previous commit

mikedlr committed:
rds_instance module - internally consistent error messages via fail_json in case of AWS exceptions

michael-dev2rights commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -14,6 +14,9 @@
 # along with Ansible. If not, see <http://www.gnu.org/licenses/>.
 
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
```
I had a look at q;  it's fine for debugging and I'll probably use it somehow, but it seems that it could leave debugging information in a public file in /tmp which could be a security problem if it accidentally went into production so I'm not totally convinced by it and I won't use it in any case in code I push up to github until I better understand how to use it securely.

mikedlr committed:
partial PEP8 cleanup - flake8 is not reporting errors for some reason
